File: lrucache.py
class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.dict = {}
        self.pqueue = [None]*capacity

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.dict.keys():
            return -1
        else:
            index = self.pqueue.index(key)
            del self.pqueue[index]
            self.pqueue.insert(0,key)
            return self.dict[key]

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        
        self.dict[key] = value
        if key not in self.pqueue:
            self.pqueue.insert(0,key)
            victim = self.pqueue.pop()
            if victim != None:
                del self.dict[victim]
        else:
            index = self.pqueue.index(key)
            del self.pqueue[index]
            self.pqueue.insert(0,key)
            

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.dict = collections.OrderedDict()
        
    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.dict:
            return -1
        value = self.dict.pop(key)
        self.dict[key] = value
        return self.dict[key]
        
    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        if key in self.dict:
            self.dict.pop(key)
        if key not in self.dict and len(self.dict) == self.capacity:
            self.dict.popitem(last=False)
        
        self.dict[key] = value

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: lrucache.cpp
class LRUCache {
public:
    LRUCache(int capacity): _capacity(capacity) {
        
    }
    
    int get(int key) {
        if (cache.find(key) == cache.end()) return -1;
        auto it = cache[key];
        int value = it->second;
        used.erase(it);
        used.push_front({key, value});
        cache[key] = used.begin();
        return value;
    }
    
    void put(int key, int value) {
        if (cache.find(key) != cache.end()) {
            auto it = cache[key];
            used.erase(it);
            used.push_front({key, value});
        }
        else {
            if (used.size() < _capacity) {
                used.push_front({key,value});
            }
            else {
                //auto it = used.end()
                auto _key = used.back().first;
                used.pop_back();
                cache.erase(_key);
                used.push_front({key,value});
            }
        }
        cache[key] = used.begin();
        
    }
private:
    unordered_map<int, list<pair<int, int> >::iterator> cache;
    list<pair<int, int> > used;
    int _capacity;
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
